Refactor device index bound check for xpu code

guangyey opened the pull request:
# Movitation
refer to [Increased compile time max GPUs to 512. Switched to int16_t DeviceIndex.](https://github.com/pytorch/pytorch/pull/119639), we use `c10::Device::MAX_NUM_DEVICES` to make sure the number of XPU devices is valid in PyTorch.

# Solution
Use `TORCH_CHECK` to check if the number of XPU devices exceeds `c10::Device::MAX_NUM_DEVICES` when enum XPU devices.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

guangyey commented on c10/xpu/XPUFunctions.cpp:
```diff
@@ -41,15 +42,19 @@ void enumDevices(std::vector<std::unique_ptr<sycl::device>>& devices) {
     }
   }
+  // We need to check the number of XPU devices to prevent DeviceIndex overflow.
+  TORCH_CHECK(
```
This function `enumDevices` could be called only once when XPU runtime is initialized. So we put the check for DeviceIndex overflow here. Meanwhile, in our design, XPU runtime doesn't depend on `c10::Device::MAX_NUM_DEVICES`. So here only check the number of XPU devices should not be greater than std::numeric_limits<DeviceIndex>::max(),

guangyey commented on c10/xpu/XPUFunctions.h:
```diff
@@ -32,4 +32,14 @@ C10_XPU_API void get_device_properties(
 C10_XPU_API int get_device_idx_from_pointer(void* ptr);
 
+static inline void check_device(DeviceIndex device) {
+  TORCH_CHECK(
```
The other component can share this code.

EikanWang commented:
@guangyey could you describe more in your pr?

tringwald commented on c10/xpu/XPUFunctions.cpp:
```diff
@@ -31,7 +31,8 @@ struct DevicePool {
 
-void enumDevices(std::vector<std::unique_ptr<sycl::device>>& devices) {
+// Ensures we only call enumDevices only once.
+void enumDevices() {
```
This could maybe be moved into a lambda function guarded by `call_once` (see my other comment below).

guangyey replied:
Change the parameter `devices` back to `enumDevices to prevent unintentional corrupting of the device list, since the `initGlobalDevicePoolState` only can be called once.

albanD commented on c10/xpu/XPUFunctions.cpp:
```diff
@@ -65,7 +75,7 @@ inline void initDevicePoolCallOnce() {
 }
 
-void initDeviceProperties(DeviceProp* device_prop, int device) {
+void initDeviceProperties(DeviceProp* device_prop, DeviceIndex device) {
```
btw if you want to find all other raw ints to remove them, regex github search is pretty effective these days! https://github.com/search?q=repo%3Apytorch%2Fpytorch+%2F%28+%7C%2C%7C%5E%29int+%2F+language%3AC%2B%2B++path%3A%2F%5E%28c10%5C%2Fxpu%5C%2F%7Caten%5C%2Fsrc%5C%2FATen%5C%2Fxpu%5C%2F%29%2F&type=code

guangyey replied:
@cyyever [[DeviceIndex][7/N] Use DeviceIndex in XPU](https://github.com/pytorch/pytorch/pull/120576#top) help us to replace other raw ints with DeviceIndex.

tringwald commented on c10/xpu/XPUFunctions.cpp:
```diff
@@ -65,7 +75,7 @@ inline void initDevicePoolCallOnce() {
   c10::call_once(init_flag, initGlobalDevicePoolState);
```
`enumDevices` should probably be guarded by `call_once` instead. Right now, it would be possible to call `initGlobalDevicePoolState` twice and corrupt the device list.

guangyey replied:
`initGlobalDevicePoolState` is guarded by `call_once`. So `initGlobalDevicePoolState` only can be called once.

tringwald replied:
That's correct for calls via `initDevicePoolCallOnce`. Someone could (by mistake) still call `enumDevices` twice. But it should not be a problem if we take care to only call `initDevicePoolCallOnce`.

guangyey commented:
Don't merge this PR. It depends on [#119639](https://github.com/pytorch/pytorch/pull/119639).

guangyey commented:
Looks like https://github.com/pytorch/pytorch/pull/119639 is stale. I will rebase to the main branch and land it.

guangyey commented:
@pytorchbot merge

guangyey committed:
refactor check device overflow for xpu code

ghstack-source-id: 1863aaf7bee13307c7b8393825805e2e23cc58f2
Pull Request resolved: https://github.com/pytorch/pytorch/pull/140369

guangyey committed:
avoid to corrupt device list

ghstack-source-id: c8c2e4a947fb2b978cdb773e07cdfa0c2950ce26
Pull Request resolved: https://github.com/pytorch/pytorch/pull/140370
